sbndqm/Decode/PMT/DaqDecoderIcarusPMT_module.cc

```diff
@@ -0,0 +1,238 @@
+  size_t const eff_fragment_id = fragment_id & 0x0fff;
+
+  sbndaq::CAENV1730Fragment fragment(artdaqFragment);
+  sbndaq::CAENV1730FragmentMetadata metafrag = *fragment.Metadata();
+  sbndaq::CAENV1730Event evt = *fragment.Event();
+  sbndaq::CAENV1730EventHeader header = evt.Header;
+  size_t nChannelsPerBoard = metafrag.nChannels;
+
+  
+  uint32_t ev_size_quad_bytes         = header.eventSize;
+  uint32_t evt_header_size_quad_bytes = sizeof(sbndaq::CAENV1730EventHeader)/sizeof(uint32_t);
+  uint32_t data_size_double_bytes     = 2*(ev_size_quad_bytes - evt_header_size_quad_bytes);
+  uint32_t nSamplesPerChannel         = data_size_double_bytes/nChannelsPerBoard;
+  uint16_t const enabledChannels      = header.ChannelMask();
+
+  artdaq::Fragment::timestamp_t const fragmentTimestamp = artdaqFragment.timestamp(); 
+  unsigned int const time_tag =  header.triggerTimeTag;
+
+  auto const [ chDataMap, nEnabledChannels ] = setBitIndices<16U>(enabledChannels);
+  const uint16_t* data_begin = reinterpret_cast<const uint16_t*>(artdaqFragment.dataBeginBytes() + sizeof(sbndaq::CAENV1730EventHeader));
+
+  std::vector<uint16_t> wvfm(nSamplesPerChannel);
+  float temperature = 0;
+
+
+  for( size_t digitizerChannel=0; digitizerChannel<nChannelsPerBoard; digitizerChannel++ ) {
+
+    std::size_t const pmtID = digitizerChannel+nChannelsPerBoard*eff_fragment_id;
+
+    std::size_t const channelPosInData = chDataMap[digitizerChannel];
+    std::size_t const ch_offset = channelPosInData * nSamplesPerChannel;
+
+    std::copy_n(data_begin + ch_offset, nSamplesPerChannel, wvfm.begin());
+
+    fOpDetWaveformCollection->emplace_back(fragmentTimestamp, pmtID, wvfm);
+
+    temperature += float( metafrag.chTemps[digitizerChannel] );
+
+  }
+
+  temperature /= nChannelsPerBoard;
+
+  fPMTDigitizerInfoCollection->emplace_back( eff_fragment_id, time_tag, fragmentTimestamp, temperature );
+
+}
+
+
+void daq::DaqDecoderIcarusPMT::produce(art::Event & event)
+{
+
+  // Make the list of the input fragments 
+
+  try {
+
+    auto fragmentHandles = readHandles( event );
+
+    auto fragments = readFragments( fragmentHandles );
+
+    // initialize the data product 
+    fOpDetWaveformCollection = OpDetWaveformCollectionPtr(new OpDetWaveformCollection);
+    fPMTDigitizerInfoCollection = PMTDigitizerInfoCollectionPtr(new PMTDigitizerInfoCollection);
+    
+    if ( fragments.size() > 0 ){
+ 
+      for( auto const & fragment : fragments ) {   
+  
+          processFragment( fragment );
+   
+      }
+    
+    }
+    
+    else {
+
+       mf::LogError("DaqDecoderIcarus") 
+	 << "No fragments found\n" << '\n';
+    }
+    
+
+    // Place the data product in the event stream
+    event.put(std::move(fOpDetWaveformCollection));
+    event.put(std::move(fPMTDigitizerInfoCollection));
+
+  }
+
+  catch( cet::exception const& e ){
+
+    mf::LogError("DaqDecoderIcarus") 
+      << "Error while attempting to decode PMT data:\n" << e.what() << '\n';
+
+  }
+
```

Comment:
To be sure that nothing escapes here, I'd add also:
```cpp
  catch(std::exception const& e) {
    mf::LogError("DaqDecoderIcarus")
      << "Exception of type " << typeid(e).name() << " caught:\n" << e.what();
  }
  catch(...) {
    mf::LogError("DaqDecoderIcarus")
      << "Non-standard exception caught. Ignoring it.";
  }
```
(the use of `typeid()` requires `#include <typeinfo>`)